### ISSUE TITLE
Error when using Compressed SpriteFonts in Pipeline

### ISSUE THREAD

**Issue by dellis1972** [mono/MonoGame#4156]
@KonajuGames I came across this when testing on the mac.. 

steps to repo

1) create a new content project
2) make sure Platform is set to Windows
3) add a sprite font
4) set sprite font TextureFormat to be "Compressed" 
5) Build

System.InvalidOperationException: Could not copy between Color and Dxt3
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent.Copy (Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent sourceBitmap, Rectangle sourceRegion, Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent destinationBitmap, Rectangle destinationRegion) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent.Copy (Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent sourceBitmap, Microsoft.Xna.Framework.Content.Pipeline.Graphics.BitmapContent destinationBitmap) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent.ConvertBitmapType (System.Type newBitmapType) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.Compress (System.Type targetType, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Boolean generateMipMaps) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressDxt (GraphicsProfile profile, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, Boolean generateMipMaps, Boolean premultipliedAlpha, Boolean sharpAlpha) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Graphics.GraphicsUtil.CompressTexture (GraphicsProfile profile, Microsoft.Xna.Framework.Content.Pipeline.Graphics.TextureContent content, TextureProcessorOutputFormat format, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context, Boolean generateMipmaps, Boolean premultipliedAlpha, Boolean sharpAlpha) in <filename unknown>:line 0
   at Microsoft.Xna.Framework.Content.Pipeline.Processors.FontDescriptionProcessor.Process (Microsoft.Xna.Framework.Content.Pipeline.Graphics.FontDescription input, Microsoft.Xna.Framework.Content.Pipeline.ContentProcessorContext context) in <filename unknown>:line 0

**Comment by cra0zy** [comment]
I noticed the same thing the day before yesterday on Linux, so it's not only Mac.

**Comment by mdhausman** [comment]
Is there any way at all to get the pipeline tool to process fonts on Linux? Or do I need to process my fonts on Windows and copy them over to my Linux dev environment?

**Comment by dellis1972** [comment]
You can use Fonts on Linux.. 
If you are using .ttf fonts just make sure the .ttf file is in the same directory as the .spritefont file.
e.g if you use Arial in your .spritefont put Arial.ttf in the same directory as the content pipeline should take care of the rest

**Comment by cra0zy** [comment]
Just use "Color" for TextureFormat instead of "Compressed"

**Comment by cra0zy** [comment]
![helpimage](https://cloud.githubusercontent.com/assets/6773302/11313332/6805e18a-8fdc-11e5-94b4-410c94493227.png)

**Comment by mdhausman** [comment]
My hero!

**Comment by mdhausman** [comment]
... Actually, I spoke too soon. It now builds without error, but it still tries to build from the /usr/share version, and MonoDevelop still doesn't like the resulting XNB.

Copied Arial.ttf to the same directory as the .spritefont file per the advice of @dellis1972. Building still results in the following:

    Cleaning /home/[snip]/Content/bin/DesktopGL/fonts/Arial12.xnb
    /home/[snip]/Content/fonts/Arial12.spritefont
    Building Font /usr/share/fonts/truetype/msttcorefonts/Arial.ttf

    Build 1 succeeded, 0 failed.

I added the resulting Arial12.xnb to my project Content, ensured that it's placed appropriately in the output folder. From the .csproj:

``` xml
   <Content Include="Content\bin\DesktopGL\fonts\Arial12.xnb">
      <Link>Content\fontXNB\Arial12.xnb</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
   </Content>
```

I try to load it with  `font = Content.Load<SpriteFont> ("fontXNB/Arial12");` and the result is a `ContentLoadException: Could not load fontXNB/Arial12 asset as a non-content file!` I got this exception earlier for a sprite that wasn't output to the right place, but I've double checked and this file is definitely in the right place, at [solution]/[project]/bin/Debug/Content/fontXNB/Arial12.xnb.

I apologize, I know that this isn't really a place for newbies to come looking for solutions, but I can't find much in the way of recent resources for MonoGame's Linux devs.

**Comment by dellis1972** [comment]
@mdhausman its ok if it picks is up from /usr/share/fonts :) The main reason to place it in the same folder is just in case you had not installed it.. but you had :) 

You're building for DesktopGL which is good. Next which version of the MonoGame.Framework are you using? If its an older version it won't be able to load the DesktopGL content.. ( @cra0zy correct? )

**Comment by mdhausman** [comment]
From packages.config:

``` xml
<packages>
  <package id="MonoGame.Framework.Linux" version="3.4.0.459" targetFramework="net45" />
</packages>
```

**Comment by cra0zy** [comment]
> If its an older version it won't be able to load the DesktopGL content.. ( @cra0zy correct? )

Correct

> And there are a lot of instances of Skipping 'OpenTK.dll.config' because it was modified and Skipping 'Tao.Sdl.dll.config' because it was modified. Which... I have definitely not modified either of those files; I've never even opened them. I'm not sure if that's related or not.

Those files are meant for waypoint to correct system libraries, ie. opengl32.dll is on Windows, libGL.so.1 is Linux and OpenGL is Mac. It's a text file.

Now for the solution to your problem, you either have to use MonoGame 3.5 dll files (if you installed the development version they are located at: "/usr/lib/mono/xbuild/MonoGame/v3.0/Assemblies/Linux") or use MonoGame 3.4 Pipeline tool.

**Comment by mdhausman** [comment]
MY HERO! For real, this time. I never got the build for 3.4 working properly, and the plug-in for MonoDevelop only covers 3.4, but I got it working by just copying the DLLs from 3.5 into the package on top of 3.4's. Success, fonts display with no trouble at all.

Thank you both so much.

**Comment by KonajuGames** [comment]
I think Dean's original issue might be a Mono on Linux codegen issue I discovered here.  Creating a new Rectangle with parameters from the calling function as the parameters to the Rectangle constructor resulted in the constructor's height parameter getting a garbage value.  Assigning the height parameter to a local variable and passing that local variable as the height to the Rectangle constructor fixed it.

**Comment by dellis1972** [comment]
@KonajuGames I'll re test this later this evening with the latest unstable and see if I can still replicate. if not i'll close this issue :)

**Comment by dellis1972** [comment]
@KonajuGames still have the same issue :(

**Comment by KonajuGames** [comment]
In MonoGame.Framework.Content.Pipeline/Graphics/DxtBitmapContent.cs, try
changing line 60 from

    if (destinationRegion != new Rectangle(0, 0, Width, Height))


to

    var h = Height;    if (destinationRegion != new Rectangle(0, 0, Width, h))

​

**Comment by cra0zy** [comment]
@KonajuGames Tested it on Linux, does not work :(

**Comment by dellis1972** [comment]
@KonajuGames on Mac its the call to PVRTexture.Transcode that raises the error.

**Comment by dellis1972** [comment]
@KonajuGames that code change didn't work. The texture in question is 512x256 not sure if that has any  impact on the issue.

**Comment by dellis1972** [comment]
@KonajuGames the Rect sizes all look ok. The source bitmap is a Vector4

**Comment by dellis1972** [comment]
@KonajuGames I wonder if this is a bit in the PVRTexture code...

**Comment by dellis1972** [comment]
@tomspilman @KonajuGames I don't think we can release 3.5 until this one is fixed as its a regression from the current stable.

**Comment by tomspilman** [comment]
@dellis1972 - Agreed.

**Comment by dellis1972** [comment]
@KonajuGames @tomspilman PVRTC via PVRLib works on Mac, is it possible that DXT compression has not been included in the mac build?

**Comment by dellis1972** [comment]
Have confirmed that ETC1, PVRTC, 16bit, ATC and color all work, its just down to DXT. DXT3 DXT5 both fail with the same problem.

**Comment by cra0zy** [comment]
Why are ETC1, PVRTC and ATC not supported on Linux?

**Comment by KonajuGames** [comment]
ATC requires a separate DLL obtained from elsewhere (we cannot distribute it).  ETC1 and PVRTC should work.  I had done some here on Linux when doing the big texture processing update.  The issue with DXT might be in the PVRTexLib libraries themselves.  I think I'll have to go back to the Nvidia libraries for DXT as PVRTexLib seems to have inconsistent support for the format across different platforms, from what I have found searching forums.

**Comment by tomspilman** [comment]
> PVRTexLib seems to have inconsistent support for the format across different platforms

Ugh...  when we were sold this idea of moving to `PVRTexLib` it was that it supported all the formats and was cross platform.  This just puts us back where we were.

**Comment by dellis1972** [comment]
Well at least its just nvidia for dxt everything else is covered

**Comment by cra0zy** [comment]
> ETC1 and PVRTC should work.

Probably should mention, they don't :(

**Comment by dellis1972** [comment]
The work on Mac :) So thats a start.
What errors do you get on Linux?

**Comment by cra0zy** [comment]
```
Could not convert texture. System.PlatformNotSupportedException: Etc1Compressed platform only supports DXT texture compression
```

**Comment by dellis1972** [comment]
@cra0zy thats because you are building for a desktop platform :) Try changing to Android and it should work.. save for PVRTC that will only work for iOS and Android. Desktop graphics hardware generally only supports DXT and its variants.

**Comment by dellis1972** [comment]
save for PVRTC =same for PVRTC ;)

**Comment by cra0zy** [comment]
> @cra0zy thats because you are building for a desktop platform :) Try changing to Android and it should work.. save for PVRTC that will only work for iOS and Android. Desktop graphics hardware generally only supports DXT and its variants.

You are right.

**Comment by KonajuGames** [comment]
> Could not convert texture. System.PlatformNotSupportedException: Etc1Compressed platform only supports DXT texture compression

And I'll fix that error message too.  `Etc1Compressed` is not a platform.

**Comment by KonajuGames** [comment]
I've been working on this tonight.  Just can't get PVRTexLib to build a DXT texture on Linux.  Found some other issues along the way as well which have been fixed (checking alpha values was incorrect because it was sometimes fed byte data from a Vector4 bitmap rather than a Color bitmap).  Tomorrow night I'll change it to go back to NVTT for DXT.  Glad I kept that code in there for just this occasion.

Also, I have to confirm if PVRTexLib contains 64-bit support on Mac OS X, which should be needed when building from Xamarin.Mac.  They distribute a dylib labelled x86, but I've yet to see if it is a universal binary.  Not that I can test with Xamarin.Mac yet though.

**Comment by dellis1972** [comment]
@KonajuGames how is this going? Need any help?

**Comment by cra0zy** [comment]
Where is PVRTexLib source?

**Comment by cra0zy** [comment]
PVRTexture.Transcode cause the crash on Linux.

**Comment by prollin** [comment]
@cra0zy I believe it was taken from [here](https://github.com/flyingdevelopmentstudio/PVRTexLibNET). Monogame's version might have been modified though and I am not sure where the modified sources are.

**Comment by dellis1972** [comment]
Isn't it from https://github.com/SickheadGames/ManagedPVRTC ?

**Comment by cra0zy** [comment]
https://github.com/KonajuGames/PVRTexLibNET

**Comment by cra0zy** [comment]
I was interested about native lib, not the C# wrapper.

**Comment by prollin** [comment]
https://community.imgtec.com/developers/powervr/installers/

**Comment by KonajuGames** [comment]
PVRTexLibNET has two parts
- PVRTexLibNET which is a C# binding for
- PVRTexLibWrapper, a C++ DLL/shared library/dylib that wraps PVRTexLib.

**Comment by KonajuGames** [comment]
I just submitted #4496 which reverts to NVTT for DXT,  It is a lot slower, but it doesn't crash on Mac and Linux.  I tried for a long time to find the cause of the crash on Linux, but it is inside PVRTexLib.  I also tried using libsquish directly.  It is what NVTT uses internally, and it was a lot faster and simpler than NVTT, but several blocks in the compressed output were incorrect.  So unfortunately I had to go back to NVTT.  I would like to investigate libsquish further as I had successfully modified it to take a BGRA Vector4 format image directly without having to go through RGBA Color like we do with NVTT.  And it does the entire compression in native code rather than calling back to C# for every compressed block.

**Comment by KonajuGames** [comment]
And we could also modify libsquish to do the specialized DXT3 compression for font textures that is mentioned in #2853.

**Comment by dellis1972** [comment]
@KonajuGames do you think we can add a MGCB unit test for that? so we can
check future regressions on Mac at least. Although I don't think the tests
run on mac atm.

**Comment by cra0zy** [comment]
This can be closed since it's been fixed with #4496
